ipa-core/src/protocol/basics/check_zero.rs

```diff
@@ -124,4 +187,122 @@ mod tests {
 
         Ok(())
     }
+
+    pub struct NotifyOnceCell<T: Clone + Send + Sync> {
+        inner: Mutex<NotifyOnceCellInner<T>>,
+    }
+
+    struct NotifyOnceCellInner<T: Clone + Send + Sync> {
+        cell: OnceCell<T>,
+        notify: Arc<Notify>,
+    }
+
+    impl<T: Clone + Send + Sync> NotifyOnceCell<T> {
+        pub fn new() -> Self {
+            Self {
+                inner: Mutex::new(NotifyOnceCellInner {
+                    cell: OnceCell::new(),
+                    notify: Arc::new(Notify::new()),
+                }),
+            }
+        }
+
+        pub fn set(&self, value: T) -> Result<(), T> {
+            let inner = self.inner.lock().unwrap();
+            inner.cell.set(value)?;
+            inner.notify.notify_waiters();
+            Ok(())
+        }
+
+        pub fn get(&self) -> Pin<Box<dyn Future<Output = T> + Send + '_>> {
+            let inner = self.inner.lock().unwrap();
+            if let Some(value) = inner.cell.get() {
+                return ready(value.clone()).boxed();
+            }
+            let notify = inner.notify.clone();
+            async move {
+                notify.notified().await;
+                self.inner.lock().unwrap().cell.get().unwrap().clone()
+            }
+            .boxed()
+        }
+    }
+
+    struct MaliciousCheckZeroInterceptor {
+        sh2: NotifyOnceCell<Fp32BitPrime>,
+    }
+
+    impl MaliciousCheckZeroInterceptor {
+        fn new() -> Self {
+            Self {
+                sh2: NotifyOnceCell::new(),
+            }
+        }
+    }
+
+    impl StreamInterceptor for MaliciousCheckZeroInterceptor {
+        type Context = InspectContext;
+
+        fn peek<'a>(
+            &'a self,
+            ctx: &'a Self::Context,
+            data: &'a mut Vec<u8>,
+        ) -> Pin<Box<dyn Future<Output = ()> + Send + 'a>> {
+            if ctx
+                .gate
+                .as_ref()
+                .contains(super::Step::MultiplyWithR.as_ref())
+                && ctx.identity == HelperIdentity::ONE
+                || ctx.gate.as_ref().contains(super::Step::RevealR.as_ref())
+                    && ctx.identity == HelperIdentity::ONE
+                    && ctx.dest == HelperIdentity::TWO.as_str()
+            {
+                async {
+                    assert_eq!(data.len(), 4);
+                    let (sh1, sh2) = join(super::SH1.get(), self.sh2.get()).await;
+                    tracing::info!("got shares: {sh1:?} {sh2:?}");
+                    let adjusted_share = -sh1 - sh2;
+                    tracing::info!("adjusted share {adjusted_share:?}");
+                    adjusted_share.serialize(
```

Comment:
I've added `serialize_to_slice` and `deserialize_from_slice` methods to `Serializable` trait (only for tests). I think you may want to use them here